pkg/zone/available_services.go

```diff
@@ -0,0 +1,166 @@
+	logger logr.Logger,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	interval time.Duration,
+	ingressTagFilters []string,
+	zone string,
+) (*ZoneAvailableServicesTracker, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_zone_available_services",
+		Help:       "Summary of available services tracker component interval",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+	return &ZoneAvailableServicesTracker{
+		logger:            logger,
+		metric:            metric,
+		resManager:        resManager,
+		interval:          interval,
+		ingressTagFilters: ingressTagFilters,
+		zone:              zone,
+	}, nil
+}
+
+func (t *ZoneAvailableServicesTracker) Start(stop <-chan struct{}) error {
+	t.logger.Info("starting")
+	ticker := time.NewTicker(t.interval)
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	for {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			if err := t.updateZoneIngresses(ctx); err != nil {
+				t.logger.Error(err, "couldn't update ZoneIngress available services")
+			}
+			t.metric.Observe(float64(time.Since(start).Milliseconds()))
+		case <-stop:
+			t.logger.Info("stopping")
+			return nil
+		}
+	}
+}
+
+func (t *ZoneAvailableServicesTracker) NeedLeaderElection() bool {
+	return true
+}
+
+func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService, other []*mesh_proto.ZoneIngress_AvailableService) bool {
+	if len(services) != len(other) {
+		return false
+	}
+	for i := range services {
+		if !proto.Equal(services[i], other[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func (t *ZoneAvailableServicesTracker) getIngressExternalServices(
+	ctx context.Context,
+	meshes []*core_mesh.MeshResource,
+) (core_mesh.ExternalServiceResourceList, error) {
+	var externalServices []*core_mesh.ExternalServiceResource
+	allMeshExternalServices := core_mesh.ExternalServiceResourceList{}
+
+	for _, mesh := range meshes {
+		if !mesh.ZoneEgressEnabled() {
+			continue
+		}
+
+		ess := core_mesh.ExternalServiceResourceList{}
+		if err := t.resManager.List(ctx, &ess, store.ListByMesh(mesh.GetMeta().GetName())); err != nil {
+			return core_mesh.ExternalServiceResourceList{}, err
+		}
+
+		// look for external services that are only available in my zone and expose them
+		for _, es := range ess.Items {
+			if es.Spec.Tags[mesh_proto.ZoneTag] == t.zone {
+				externalServices = append(externalServices, es)
+			}
+		}
+	}
+
+	allMeshExternalServices.Items = externalServices
+	return allMeshExternalServices, nil
+}
+
+func (t *ZoneAvailableServicesTracker) updateZoneIngresses(ctx context.Context) error {
+	meshes := core_mesh.MeshResourceList{}
+	if err := t.resManager.List(ctx, &meshes); err != nil {
+		return err
+	}
+
+	dps := core_mesh.DataplaneResourceList{}
+	if err := t.resManager.List(ctx, &dps); err != nil {
+		return err
+	}
```

Comment:
This call concerns me when it comes to perf. What I read here is listing DPPs using uncached resource manager every second. What we could do here is

1) Use cached resource manager and bump default interval to maybe something like 5s
2) Rewrite this to event based approach but I don't think I'd bother since long term this will be gone
3) Use `MeshCache.GetMesh()`, this way we can leverage resources that were already fetched for the mesh. We get all the resources for free in O(1) assuming that there is at least one DPP connected of mesh X connected to this CP

Reply:
Went with option 3.